clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -7042,12 +7098,44 @@ void SPIRVTranslator::ConstructJob(Compilation &C, const JobAction &JA,
     TranslatorArgs.push_back("-spirv-ext=+all");
   }
   for (auto I : Inputs) {
-    TranslatorArgs.push_back(I.getFilename());
+    std::string Filename(I.getFilename());
+    if (I.getType() == types::TY_TempIRfilelist) {
+      ForeachArgs.push_back(
+          C.getArgs().MakeArgString("--in-file-list=" + Filename));
+      ForeachArgs.push_back(
+          C.getArgs().MakeArgString("--in-replace=" + Filename));
+      ForeachArgs.push_back(
+          C.getArgs().MakeArgString("--out-ext=spv"));
+    }
+    TranslatorArgs.push_back(C.getArgs().MakeArgString(Filename));
   }
 
-  C.addCommand(std::make_unique<Command>(JA, *this,
+  auto Cmd = std::make_unique<Command>(JA, *this,
       TCArgs.MakeArgString(getToolChain().GetProgramPath(getShortName())),
-      TranslatorArgs, None));
+      TranslatorArgs, None);
+
+  if (!ForeachArgs.empty()) {
+    // Construct llvm-foreach command.
+    // The llvm-foreach command looks like this:
+    // llvm-foreach a.list --out-replace=out "cp {} out"
+    // --out-file-list=list
+    std::string OutputFileName(Output.getFilename());
+    ForeachArgs.push_back(
+        TCArgs.MakeArgString("--out-file-list=" + OutputFileName));
+    ForeachArgs.push_back(
+        TCArgs.MakeArgString("--out-replace=" + OutputFileName));
+    ForeachArgs.push_back(TCArgs.MakeArgString("--"));
+    ForeachArgs.push_back(TCArgs.MakeArgString(Cmd->getExecutable()));
+
+    for (auto &Arg : Cmd->getArguments())
+      ForeachArgs.push_back(Arg);
+
+    SmallString<128> ForeachPath(C.getDriver().Dir);
+    llvm::sys::path::append(ForeachPath, "llvm-foreach");
+    const char *Foreach = C.getArgs().MakeArgString(ForeachPath);
+    C.addCommand(std::make_unique<Command>(JA, *this, Foreach, ForeachArgs, None));
+  } else
+    C.addCommand(std::move(Cmd));
```

Comment:
I understand you don't reuse `constructLLVMForeachCommand()` here because the function resides in `SYCL::` toolchains area, whereas the Translator is a common tool. However, I'd argue that there're no SYCL specifics to `constructLLVMForeachCommand()`, so maybe "globalizing" it and reusing for llvm-spirv, llvm-link would be better.

Reply:
Ok, I'll see what I can do for it. I remember that place for "globalizing" `constructLLVMForeachCommand` wasn't obvious for me ( but I assume it's happened because I'm a new person in the driver), so it would be nice if you have a suggestion.

Reply:
I was wrong by saying that "constructLLVMSPIRV" is common, and on the whole, you can probably reuse `constructLLVMForeachCommand` w/o further modification.

Reply:
Not sure that I got what you meant.. I'm a new person in the driver, so, could you please explain why "common" function hould be considered such if defined in `ToolChains/Gnu.h`?

Reply:
Please disregard. I'm new enough to Driver to make wrong assumptions here and there :)
Again, the function should be safely reusable as is.

Reply:
Do you mean that I can duplicate the function for both toolchains?

Reply:
Checked offline that defining the function for the Clang toolchain and including it for the SYCL toolchain doesn't work. With that, I'm fine if the Clang.cpp part doesn't reuse the function.